Plugin crash on abstract model filtering with custom Manager

python 3.10
django 4.1.5
mypy 0.991
django-stubs 0.14.0
django-stubs-ext 0.7.0

This is a re-issue of #1189, with a fixed example. #1189 was erroneously closed (because the example given wasn't reproducing the issue properly).

Crash in plugin when using an abstract model with a custom manager:

```
./myproj/app/models.py:15: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 0.982
Traceback (most recent call last):
  File "mypy/checkexpr.py", line 4619, in accept
  File "mypy/nodes.py", line 1822, in accept
  File "mypy/checkexpr.py", line 385, in visit_call_expr
  File "mypy/checkexpr.py", line 505, in visit_call_expr_inner
  File "mypy/checkexpr.py", line 1155, in check_call_expr_with_callee_type
  File "mypy/checkexpr.py", line 1238, in check_call
  File "mypy/checkexpr.py", line 1438, in check_callable_call
  File "mypy/checkexpr.py", line 995, in apply_function_plugin
  File "django-stubs/mypy_django_plugin/transformers/orm_lookups.py", line 40, in typecheck_queryset_filter
    lookup_type = django_context.resolve_lookup_expected_type(ctx, model_cls, lookup_kwarg)
  File "django-stubs/mypy_django_plugin/django/context.py", line 388, in resolve_lookup_expected_type
    lookup_parts, field_parts, is_expression = query.solve_lookup_type(lookup)
  File "/code/myproj/.venv/lib/python3.10/site-packages/django/db/models/sql/query.py", line 1217, in solve_lookup_type
    _, field, _, lookup_parts = self.names_to_path(lookup_splitted, self.get_meta())
  File "/code/myproj/.venv/lib/python3.10/site-packages/django/db/models/sql/query.py", line 1656, in names_to_path
    name = opts.pk.name
AttributeError: 'NoneType' object has no attribute 'name'
./myproj/app/models.py:15: : note: use --pdb to drop into pdb
```

Minimal reproduction example (with or without django_stubs_ext.monkeypatch):

NB: This needs to be in a models file in an installed app.

```python
from django.db import models


class BaseManager(models.Manager):
    pass


class BaseModel(models.Model):
    objects = BaseManager()

    class Meta:
        abstract = True

    def lock(self) -> None:
        type(self).objects.filter(pk=self.pk).select_for_update().values_list("pk")


class MyModel(BaseModel):
    pass
```

--- a/mypy_django_plugin/django/context.py
+++ b/mypy_django_plugin/django/context.py
@@ -2,11 +2,12 @@ import os
 import sys
 from collections import defaultdict
 from contextlib import contextmanager
-from typing import TYPE_CHECKING, Any, Dict, Iterable, Iterator, Optional, Set, Tuple, Type, Union
+from typing import TYPE_CHECKING, Any, Dict, Iterable, Iterator, Optional, Sequence, Set, Tuple, Type, Union
 
 from django.core.exceptions import FieldError
 from django.db import models
 from django.db.models.base import Model
+from django.db.models.expressions import Expression
 from django.db.models.fields import AutoField, CharField, Field
 from django.db.models.fields.related import ForeignKey, RelatedField
 from django.db.models.fields.reverse_related import ForeignObjectRel
@@ -19,6 +20,7 @@ from mypy.plugin import MethodContext
 from mypy.types import AnyType, Instance
 from mypy.types import Type as MypyType
 from mypy.types import TypeOfAny, UnionType
+from typing_extensions import Literal
 
 from mypy_django_plugin.lib import fullnames, helpers
 from mypy_django_plugin.lib.fullnames import WITH_ANNOTATIONS_FULLNAME
@@ -375,29 +377,39 @@ class DjangoContext:
         assert isinstance(field, (Field, ForeignObjectRel))
         return field
 
+    def solve_lookup_type(
+        self, model_cls: Type[Model], lookup: str
+    ) -> Optional[Tuple[Sequence[str], Sequence[str], Union[Expression, Literal[False]]]]:
+        query = Query(model_cls)
+        if (lookup == "pk" or lookup.startswith("pk__")) and query.get_meta().pk is None:
+            # Primary key lookup when no primary key field is found, model is presumably
+            # abstract and we can't say anything about 'pk'.
+            return None
+        return query.solve_lookup_type(lookup)
+
     def resolve_lookup_into_field(
         self, model_cls: Type[Model], lookup: str
-    ) -> Union["Field[Any, Any]", ForeignObjectRel]:
-        query = Query(model_cls)
-        lookup_parts, field_parts, is_expression = query.solve_lookup_type(lookup)
-
+    ) -> Union["Field[Any, Any]", ForeignObjectRel, None]:
+        solved_lookup = self.solve_lookup_type(model_cls, lookup)
+        if solved_lookup is None:
+            return None
+        lookup_parts, field_parts, is_expression = solved_lookup
         if lookup_parts:
             raise LookupsAreUnsupported()
         return self._resolve_field_from_parts(field_parts, model_cls)
 
     def resolve_lookup_expected_type(self, ctx: MethodContext, model_cls: Type[Model], lookup: str) -> MypyType:
-        query = Query(model_cls)
-        if lookup == "pk" or lookup.startswith("pk__") and query.get_meta().pk is None:
-            # Primary key lookup when no primary key field is found, model is presumably
-            # abstract and we can't say anything about 'pk'.
-            return AnyType(TypeOfAny.implementation_artifact)
         try:
-            lookup_parts, field_parts, is_expression = query.solve_lookup_type(lookup)
-            if is_expression:
-                return AnyType(TypeOfAny.explicit)
+            solved_lookup = self.solve_lookup_type(model_cls, lookup)
         except FieldError as exc:
             ctx.api.fail(exc.args[0], ctx.context)
             return AnyType(TypeOfAny.from_error)
+
+        if solved_lookup is None:
+            return AnyType(TypeOfAny.implementation_artifact)
+        lookup_parts, field_parts, is_expression = solved_lookup
+        if is_expression:
+            return AnyType(TypeOfAny.explicit)
 
         field = self._resolve_field_from_parts(field_parts, model_cls)
 

--- a/mypy_django_plugin/transformers/querysets.py
+++ b/mypy_django_plugin/transformers/querysets.py
@@ -58,7 +58,9 @@ def get_field_type_from_lookup(
     except LookupsAreUnsupported:
         return AnyType(TypeOfAny.explicit)
 
-    if (isinstance(lookup_field, RelatedField) and lookup_field.column == lookup) or isinstance(
+    if lookup_field is None:
+        return AnyType(TypeOfAny.implementation_artifact)
+    elif (isinstance(lookup_field, RelatedField) and lookup_field.column == lookup) or isinstance(
         lookup_field, ForeignObjectRel
     ):
         related_model_cls = django_context.get_field_related_model_cls(lookup_field)

--- a/tests/typecheck/models/test_abstract.yml
+++ b/tests/typecheck/models/test_abstract.yml
@@ -7,3 +7,29 @@
     AbstractUser.objects.get(pkey=1)  # ER: Cannot resolve keyword 'pkey' into field..*
   installed_apps:
     - django.contrib.auth
+
+- case: test_fetch_pk_with_custom_manager_on_abstract_model
+  main: |
+    from myapp.models import MyModel
+  installed_apps:
+    - myapp
+  files:
+    -   path: myapp/__init__.py
+    -   path: myapp/models.py
+        content: |
+            from django.db import models
+
+            class BaseManager(models.Manager):
+                pass
+
+            class BaseModel(models.Model):
+                objects = BaseManager()
+
+                class Meta:
+                    abstract = True
+
+                def lock(self) -> None:
+                    reveal_type(type(self).objects.values_list("pk"))  # N: Revealed type is "django.db.models.query._QuerySet[myapp.models.BaseModel, Tuple[Any]]"
+
+            class MyModel(BaseModel):
+                field = models.IntegerField()

